pkg/cache/internal/selector.go

```diff
@@ -0,0 +1,37 @@
+package internal
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// SelectorsByGVK associate a GroupVersionKind to a field/label selector
+type SelectorsByGVK map[schema.GroupVersionKind]Selectors
+
+// Selector is specialized client.ListOption to distinguish from other ListOptions
+// that cannot be applied to the cache's ListWatch.
+type Selector interface {
```

Comment:
Lets update the godoc to mention that only LabelSelectors and FIeldSelectors will be respected.

Reply:
And honestly, I find this much worse than what was in the original proposal. We now have an interface with an `ApplyToList` but we will only ever respect two of the possible list options and on top of that an `IsSelector` that doesn't do anything just to differentiate it from client.ListOptions.

This is so much easier to missunderstand and missuse than the original approach that was in the design doc of a simple struct with two fields: https://github.com/kubernetes-sigs/controller-runtime/blob/4aecab5d395c264b1a3d2df7cfed21489168d4be/designs/use-selectors-at-cache.md#L114-L115

/cc @vincepri @DirectXMan12 for thoughts on that

Reply:
FWIW it's easier to make `type Selector struct` conform to an interface than it is to remove `type Selector interface` later, in case we do find this to be a point of confusion. Originally I was pro-`client.ListOptions` but I don't actually think much is gained by conformance. Also despite the fact that `type Selector interface` is internal, a library user can still implement it and create a possibly no-op or wrong concrete type to `SelectorsByObject`.

Reply:
Going back to previous implementation.